### DISCUSSION TITLE
Web / WASM embeddable or remote terminal?

### DISCUSSION THREAD

**Discussion by mlvzk** [ghostty-org/ghostty#3599] [Feature Requests, Ideas] [answered]
1. Will libghostty make it easy to embed ghostty on the web? Is WASM a supported (or planned) target for libghostty?
2. Theoretically, could VS Code replace their terminal (which needs to support browsers, vscode.dev) with Ghostty?
3. Thirdly, a very experimental idea but it's related to the two points above - would it make sense for Ghostty support a "remote" mode, similar to VS Code's `code serve-web` or `code tunnel`? These commands give you a link you can open in a browser to use VS Code as a client in your browser, with the server part running on the computer where `code tunnel` was ran on.
If you wanted to implement this outside Ghostty's codebase, you would need a sort of generic 'Web SSH Client', only it would need to operate over http(s) not ssh, so you still need to run a command on some server. This 'Web SSH Client' could be implemented with Ghostty (libghostty?), if it was embeddable on the web. I think I like the external idea more, but leaving this in because I would like for this use-case to be considered.

**Comment by 00-kat** [comment, marked as answer]
Yes it's planned.

Mitchell was even thinking of some crazy ideas like using a WASM `libghostty` for the frontpage of the Ghostty website as well.

@ gabydd (intentionally avoiding a ping) was messing around with getting a WASM backend to work over in the thread at https://discord.com/channels/1005603569187160125/1309049042092691487 (it's in the [Ghostty Discord server](https://discord.com/ghostty)).

Point 2: yes; and in one of the talks somewhere (I think Zig Showtime), Mitchell mentioned that an ambitious goal of his was to replace `xterm.js` with `libghostty` in VSCode.

**Reply by mlvzk** [reply]
> Point 2: yes; and in one of the talks somewhere (I think Zig Showtime), Mitchell mentioned that an ambitious goal of his was to replace xterm.js with libghostty in VSCode.

Found it, thanks: https://youtu.be/l_qY2p0OH9A?t=4633

I also found this issue in VS Code's repo: https://github.com/microsoft/vscode/issues/236991

**Comment by gabydd** [comment]
for one if you are very excited to try it out I just got https://github.com/gabydd/ghostty/tree/wasi to work properly for 0.13

steps to make it work is:
1. checkout the branch
2. ``zig build wasm``
3. ``cd example``
4. ``npm run start``

the termio is a bit broken so sometimes bytes get eaten up and the terminal looks a bit broken

also this might not be the direction we take wasm in the future just an experiment

**Reply by mlvzk** [reply]
I got it running albeit with a lot of errors in console and graphics look a little weird, but still very cool! Thanks for sharing

**Reply by gabydd** [reply]
Errors in console should just be logging, I make it use console.error so I can know the place the log comes from.
